### DISCUSSION TITLE
Discussing the Hono Middleware for Azure Functions Custom Handler (on Nammatham v3)

### DISCUSSION THREAD

**Discussion by mildronize** [honojs/hono#2674] [General]
Dear all who developing Hono Framework, I've quiet new in Hono, I would like to dicuss the design middleware or has alternative for more cleaner code and type-safety

During design and developing [Nammatham v3](https://github.com/thaitype/nammatham/wiki/V3-Roadmap) (Azure Functions Framework), I've heard the good news the that's the [Hono support Azure Functions](https://hono.dev/getting-started/azure-functions). 

However, that Hono adapter for Azure Functions based on the [official Azure Functions v4](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/azure-functions-node-js-v4-programming-model-is-generally/ba-p/3929217) which same idea behind the Nammatham v2.

The major problem of the official Azure Functions v4 is it cannot be bundled and cannot use with the monorepo 
https://github.com/Azure/azure-functions-nodejs-library/issues/201, with Azure Custom Handler I can [build, bundling with monorepo and deploy successfully](https://github.com/thaitype/nammatham/wiki/V3-Roadmap#progress).

So, I've design the Hono Middleware for Azure Functions 
see summary design below (which support type infer from Hono):

```ts
import { Hono } from 'hono';
import { logger } from 'hono/logger';
import { FunctionTrigger } from 'nammatham';

// DO NOT SET `basePath` for Hono App, Azure Functions will handle it
const app = new Hono();
app.use(logger());

const trigger = new FunctionTrigger();

app.all(
  ...trigger.http({
    route: '/SimpleHttpTrigger',
  }),
  c => {
    // Getting the function context
    const context = c.var.func;

    context.log('JavaScript HTTP trigger function processed a request.');
    context.log(`invocationid is: ${context.invocationId}`);
    context.log(`The third log message.`);

    return context.json({
      hello: 'world',
    });
  }
);
```

the result will be 

```ts
{
    "hello": "world"
}
```

Note: This code is run with Azure Functions Host which is act like a reverse proxy forward request to Hono router

Or you can see all design interface in the [Playground](https://www.typescriptlang.org/play/?#code/PTAEAUCcHtoM1PUBjaA7ZBTADgF1HNJKABLrSgCyAlgCa0A2mA7gIaSaJqgCCAXgFcO3AGICMuaugDOoAMIDpuaAFtSrNI0yQAXKAAWuXNmk6QAc2q59AgEYA6VCuDXWVgJ7ZMwNKxUrWV2dsAQYGYABGABYAJgAoOOoVbCJ8AG9SclAAXwIYNQByfXICgG5E5NTQDIZoc3NtHLzVUCLyYFr67TKKlMh8XE9ODJINLUgAJUxpFLRpTAAaKjotNg5RzSZiXLh81uK0aBch6R7eqozkDkDMGnomNc4dvbbD4DhWZGVIdzPMAA8+vhqGhcNoPlhSEZsAAVSDULqQADyeCkcwAPDCJtABGDQACwZpZEp4WhzAA+apxUCgVi4-QAGUwADdMAwAPx6AoadDuFQ406gAA+rTg4i+aIKwtarFoKhBPRpIJCuGknNAU1QkFo6JJIPMS3EAGtDsw0OTyjSccZcWq9JqiDq9WTDWgTdAzRbqaAYLjMOqsdbMOVsglBl5MocAKJoZmgAC8VJpADV2NRWLYmKYkzSCOK9GlvbnQCDmdBkIE0QBJWh6Z3mS3FktoFXZh3a3W4UkG2lodxepugABW0nQegAFLRAqw9Bp3ABKBOUjbjKYzGSYdFzgdNzoTlTTaSsBp1rv6xfxylluiNmnZRv3uKPuIAoEoBisaSyMQSNFwhENMQhY0oYxjoqgcz4IGfr4v8hK0MSZ5kuS47QKiMh6CQ0L-oiKKSDImLYn65LznoADa0FgksdyrOwmArls6JkNGsbkgAujmNIQUooDyvcLB0QmKDXGCNEPHRTHkDGzIoZ+7gYKA47IEsaAEhelLAU23H4J0lCHse0ynt2ZEcYmJm3rmyD2PMuDjgUYoYAUSyaYOpblpW6A1noVkcAAjvY+iYLK2h2f8AC0rCCBwYUORKMhhW5Fb4WgdAFIuQoigUTlFk2yq2gW2QLDlxZ7opB5fgZRnnkunGDqAun6Q00j2CE0j6OO5VHg084WcWhXFbmI5jqAU64KwNUuXVHC4EI3BWUNaDjpNdWgEiuKtgWA0rRw2bLStuYAMpjTN0hyNAtCYHoMQAAzXUV+11QAQud7h6KNrD3Q9TaBcFkC7VtX0FBBhK4GF4aYAUXKsNg2AMNQSVosAC3ZV9Tb9ajd6fQ96NfQydTZg1FVNVj+0gBqmAzZAaCpgwAiXa0aQAESBWE0CMzojPMEQDC0Iz2Qo-t2Q9QDON9cLg6sGwVigKpcHjuLuZCxZ02zaAZFoclzW+nin6gJRiy8Ss4kcBxuvaQ+T4JHEZMACJIqAABySIwqAB1Ri7AAGtifpg4CBPoHsEEQkYUDwMNLPwQicD+cVzKAzDUGEBhjEwJa4HE2m0jDQmqcwIfy+U0PYPYiiYOOnSAfLwsZzIAzwoiOcsKAMfJThlfV0X9isGE47evY-dnoiAXQktOXa-TBTAAdSSw-R2H14BAtC59yATTlZMAOIU5IZKgNYnCxclKDoGCcE5ZnwMEkJVnMuw9ixeU58nwS9idHZABSrC3wdVzUHgpAwhhOAPeC9GiHzRKAbAMAsBfkwLQWkPpMC+TpkoewaULKXzgq-Oo44PaJQ8ileB1BswABI0iYNwPYfByUazZA9grLiz8sFvw9jCQKe99DUG1PVOovFGqYHsPQx+xYVZU2PqCF+C1R6DhZrULkXNIA8wFneBWoZq5xCAA)

**Comment by yusukebe** [comment]
cc: @Marplex

**Comment by mildronize** [comment]
Thank you so much @yusukebe, I'm really appreciate your work, extending Hono Framework is my best idea for maintaining the project which requires type-safe leveraging Hono type-inference system.

**Reply by yusukebe** [reply]
Hi! Honestly, I'm not very familiar with Azure Functions, so I'd be happy if the developers who are using Azure Functions would discuss it here!

**Comment by Marplex** [comment]
Hi! So if I understand this correctly, the standard azure functions library won't bundle with esbuild. You're bypassing this problem by not using the library at all, and using a custom handler that runs the script with Bun. Is this correct?

**Reply by mildronize** [reply]
Yes, it's correct, in V3 I don't use Azure Functions Node.js v4 anymore, just simple custom handler like go, rust lang, 
However, i need to handle a lot of stuff behind the scene.

It almost close to Nammatham v1 thats consume only type from  Azure Functions Node.js v3, it needs to handle `function.json` generation

**Comment by Marplex** [comment]
Hi! So if I understand this correctly, the standard azure functions library won't bundle with esbuild. You're bypassing this problem by not using the library at all, and using a custom handler that runs the script with Bun. Is this correct?
